src/UriSetPath.c

```diff
@@ -0,0 +1,493 @@
+			|| (uri->pathHead->text.first != uri->pathHead->text.afterLast)) {
+		return;  /* i.e. nothing to do */
+	}
+
+	assert(uri->absolutePath == URI_FALSE);
+
+	URI_FUNC(DropEmptyFirstPathSegment)(uri, memory);
+
+	uri->absolutePath = URI_TRUE;
+
+	if ((uri->pathHead == uri->pathTail) && (uri->pathHead->text.first == uri->pathHead->text.afterLast)) {
+		URI_FUNC(DropEmptyFirstPathSegment)(uri, memory);
+	}
+}
+
+
+
+static int URI_FUNC(InternalSetPath)(URI_TYPE(Uri) * destUri,
+		const URI_CHAR * first,
+		const URI_CHAR * afterLast,
+		UriMemoryManager * memory) {
+	assert(destUri != NULL);
+	assert(first != NULL);
+	assert(afterLast != NULL);
+	assert(memory != NULL);
+	assert(destUri->pathHead == NULL);  /* set by SetPathMm right before */
+	assert(destUri->pathTail == NULL);  /* set by SetPathMm right before */
+	assert(destUri->absolutePath == URI_FALSE);  /* set by SetPathMm right before */
+
+	/* Skip the leading slash from target URIs with a host (so that we can
+	 * transfer the path 1:1 further down) */
+	if (URI_FUNC(HasHost)(destUri) == URI_TRUE) {
+		/* NOTE: This is because SetPathMm called IsWellFormedPath earlier: */
+		assert((afterLast - first >= 1) && (first[0] == _UT('/')));
+		first++;
+	} else if (first == afterLast) {
+	    /* This avoids (1) all the expensive but unecessary work below
+	     * and also (2) mis-encoding as single empty path segment
+	     * that would need (detection and) repair further down oetherwise */
+	    return URI_SUCCESS;
+	}
+
+	/* Assemble "///.." input wrap for upcoming parse as a URI */
+	{
+		const size_t inputLenChars = (afterLast - first);
+		const size_t MAX_SIZE_T = (size_t)-1;
+
+		/* Detect overflow */
+		if (MAX_SIZE_T - inputLenChars < 3 + 1) {
+			return URI_ERROR_MALLOC;
+		}
+
+		{
+			const size_t candidateLenChars = 3 + inputLenChars;
+
+			/* Detect overflow */
+			if (MAX_SIZE_T / sizeof(URI_CHAR) < candidateLenChars + 1) {
+				return URI_ERROR_MALLOC;
+			}
+
+			{
+				URI_CHAR * const candidate = memory->malloc(memory, (candidateLenChars + 1) * sizeof(URI_CHAR));
+
+				if (candidate == NULL) {
+					return URI_ERROR_MALLOC;
+				}
+
+				memcpy(candidate, _UT("///"), 3 * sizeof(URI_CHAR));
+				memcpy(candidate + 3, first, inputLenChars * sizeof(URI_CHAR));
+				candidate[3 + inputLenChars] = _UT('\0');
+
+				/* Parse as an RFC 3986 URI */
+				{
+					URI_TYPE(Uri) tempUri;
+					const int res = URI_FUNC(ParseSingleUriExMm)(&tempUri,
+							candidate,
+							candidate + candidateLenChars,
+							NULL,
+							memory);
+					assert((res == URI_SUCCESS) || (res == URI_ERROR_SYNTAX) || (res == URI_ERROR_MALLOC));
+					if (res != URI_SUCCESS) {
+						memory->free(memory, candidate);
+						return res;
+					}
+
+					/* Nothing but path and host is supposed to be set by the parse, in particular not: */
+					assert(tempUri.query.first == NULL);
+					assert(tempUri.fragment.first == NULL);
+
+					/* Ensure that the strings in the path segments are all owned */
```

Comment:
I stumbled upon this and asked myself whether this is necessary, since `candidate` is already allocated by the correct MM. But it is necessary, because `candidate` is one large string and `MakeOwnerMm` splits this into multiple small strings, right?

Consider spelling this out more explicitly if I understood it correctly.

Reply:
@TimWolla improved comment coming up with next push for re-review…

Reply:
This thread is resolved.